process/timeout.rs

```diff
@@ -0,0 +1,283 @@
+    #[arg(name = "ARGUMENT", trailing_var_arg = true)]
+    arguments: Vec<String>,
+}
+
+/// Parses string slice into [Duration].
+///
+/// # Arguments
+///
+/// * s - [str] that represents input duration.
+///
+/// # Errors
+///
+/// Returns an error if passed invalid input.
+///
+/// # Returns
+///
+/// Returns [Duration].
+fn parse_duration(s: &str) -> Result<Duration, String> {
+    let (value, suffix) = s.split_at(
+        s.find(|c: char| !c.is_ascii_digit() && c != '.')
+            .unwrap_or(s.len()),
+    );
+
+    let value: f64 = value
+        .parse()
+        .map_err(|_| format!("invalid duration format '{}'", s))?;
+
+    let multiplier = match suffix {
+        "s" | "" => 1.0,
+        "m" => 60.0,
+        "h" => 3600.0,
+        "d" => 86400.0,
+        _ => return Err(format!("invalid duration format '{}'", s)),
+    };
+
+    Ok(Duration::from_secs_f64(value * multiplier))
+}
+
+/// Parses and validates the signal name, returning its integer value.
+///
+/// # Arguments
+///
+/// * s - [str] that represents the signal name.
+///
+/// # Errors
+///
+/// Returns an error if passed invalid input.
+///
+/// # Returns
+///
+/// Returns the integer value of the signal.
+fn parse_signal(s: &str) -> Result<Signal, String> {
+    let signal_name = format!("SIG{}", s.to_uppercase());
+
+    Signal::from_str(&signal_name).map_err(|_| format!("invalid signal name '{}'", s))
+}
+
+#[derive(thiserror::Error, Debug)]
+enum TimeoutError {
+    #[error("timeout reached")]
+    TimeoutReached,
+    #[error("signal sent '{0}'")]
+    SignalSent(i32),
+    #[error("{0}")]
+    Other(String),
+    #[error("unable to run the utility '{0}'")]
+    UnableToRunUtility(String),
+    #[error("utility '{0}' not found")]
+    UtilityNotFound(String),
+}
+
+impl From<std::io::Error> for TimeoutError {
+    fn from(error: std::io::Error) -> Self {
+        TimeoutError::Other(error.to_string())
+    }
+}
+
+impl From<Errno> for TimeoutError {
+    fn from(error: Errno) -> Self {
+        TimeoutError::Other(error.to_string())
+    }
+}
+
+impl From<TimeoutError> for i32 {
+    fn from(error: TimeoutError) -> Self {
+        match error {
+            TimeoutError::TimeoutReached => 124,
+            TimeoutError::SignalSent(signal) => 128 + signal,
+            TimeoutError::Other(_) => 125,
+            TimeoutError::UnableToRunUtility(_) => 126,
+            TimeoutError::UtilityNotFound(_) => 127,
+        }
+    }
+}
+
+fn send_signal(pid: Pid, signal: Signal) -> Result<(), TimeoutError> {
+    kill(pid, signal).map_err(Into::into)
+}
+
+fn wait(child: &mut Child) -> Result<i32, TimeoutError> {
+    child
+        .wait()
+        .map(ExitStatus::into_raw)
+        .map_err(Into::into)
+}
+
+fn wait_for_duration(child: &mut Child, duration: Duration) -> Result<i32, TimeoutError> {
+    drop(child.stdin.take());
+
+    let start = Instant::now();
+    while start.elapsed() < duration {
+        if let Some(status) = child
+            .try_wait()
```

Comment:
Instead of try_wait, you can use wait_timeout with crate nix to wait for the process to complete more efficiently.
If I'm wrong, correct me pls

Reply:
`nix` crate doesn't provide *wait* with timeout (https://docs.rs/nix/latest/nix/sys/wait/index.html).

There is some other crate: https://docs.rs/wait-timeout/latest/wait_timeout/. Should I include this crate into project?